File: 30_Cube/javascript/cube.js
// CUBE
//
// Converted from BASIC to Javascript by Oscar Toledo G. (nanochess)
//

function print(str)
{
    document.getElementById("output").appendChild(document.createTextNode(str));
}

function input()
{
    var input_element;
    var input_str;
    
    return new Promise(function (resolve) {
                       input_element = document.createElement("INPUT");
                       
                       print("? ");
                       input_element.setAttribute("type", "text");
                       input_element.setAttribute("length", "50");
                       document.getElementById("output").appendChild(input_element);
                       input_element.focus();
                       input_str = undefined;
                       input_element.addEventListener("keydown", function (event) {
                                                      if (event.keyCode == 13) {
                                                      input_str = input_element.value;
                                                      document.getElementById("output").removeChild(input_element);
                                                      print(input_str);
                                                      print("\n");
                                                      resolve(input_str);
                                                      }
                                                      });
                       });
}

function tab(space)
{
    var str = "";
    while (space-- > 0)
        str += " ";
    return str;
}

// Main program
async function main()
{
    print(tab(33) + "CUBE\n");
    print(tab(15) + "CREATIVE COMPUTING  MORRISTOWN, NEW JERSEY\n");
    print("\n");
    print("\n");
    print("\n");
    print("DO YOU WANT TO SEE THE INSTRUCTIONS? (YES--1,NO--0)");
    b7 = parseInt(await input());
    if (b7 != 0) {
        print("THIS IS A GAME IN WHICH YOU WILL BE PLAYING AGAINST THE\n");
        print("RANDOM DECISION OF THE COMPUTER. THE FIELD OF PLAY IS A\n");
        print("CUBE OF SIDE 3. ANY OF THE 27 LOCATIONS CAN BE DESIGNATED\n");
        print("BY INPUTING THREE NUMBERS SUCH AS 2,3,1. AT THE START,\n");
        print("YOU ARE AUTOMATICALLY AT LOCATION 1,1,1. THE OBJECT OF\n");
        print("THE GAME IS TO GET TO LOCATION 3,3,3. ONE MINOR DETAIL:\n");
        print("THE COMPUTER WILL PICK, AT RANDOM, 5 LOCATIONS AT WHICH\n");
        print("IT WILL PLANT LAND MINES. IF YOU HIT ONE OF THESE LOCATIONS\n");
        print("YOU LOSE. ONE OTHER DETAIL: YOU MAY MOVE ONLY ONE SPACE \n");
        print("IN ONE DIRECTION EACH MOVE. FOR  EXAMPLE: FROM 1,1,2 YOU\n");
        print("MAY MOVE TO 2,1,2 OR 1,1,3. YOU MAY NOT CHANGE\n");
        print("TWO OF THE NUMBERS ON THE SAME MOVE. IF YOU MAKE AN ILLEGAL\n");
        print("MOVE, YOU LOSE AND THE COMPUTER TAKES THE MONEY YOU MAY\n");
        print("HAVE BET ON THAT ROUND.\n");
        print("\n");
        print("\n");
        print("ALL YES OR NO QUESTIONS WILL BE ANSWERED BY A 1 FOR YES\n");
        print("OR A 0 (ZERO) FOR NO.\n");
        print("\n");
        print("WHEN STATING THE AMOUNT OF A WAGER, PRINT ONLY THE NUMBER\n");
        print("OF DOLLARS (EXAMPLE: 250)  YOU ARE AUTOMATICALLY STARTED WITH\n");
        print("500 DOLLARS IN YOUR ACCOUNT.\n");
        print("\n");
        print("GOOD LUCK!\n");
    }
    a1 = 500;
    while (1) {
        a = Math.floor(3 * Math.random());
        if (a == 0)
            a = 3;
        b = Math.floor(3 * Math.random());
        if (b == 0)
            b = 2;
        c = Math.floor(3 * Math.random());
        if (c == 0)
            c = 3;
        d = Math.floor(3 * Math.random());
        if (d == 0)
            d = 1;
        e = Math.floor(3 * Math.random());
        if (e == 0)
            e = 3;
        f = Math.floor(3 * Math.random());
        if (f == 0)
            f = 3;
        g = Math.floor(3 * Math.random());
        if (g == 0)
            g = 3;
        h = Math.floor(3 * Math.random());
        if (h == 0)
            h = 3;
        i = Math.floor(3 * Math.random());
        if (i == 0)
            i = 2;
        j = Math.floor(3 * Math.random());
        if (j == 0)
            j = 3;
        k = Math.floor(3 * Math.random());
        if (k == 0)
            k = 2;
        l = Math.floor(3 * Math.random());
        if (l == 0)
            l = 3;
        m = Math.floor(3 * Math.random());
        if (m == 0)
            m = 3;
        n = Math.floor(3 * Math.random());
        if (n == 0)
            n = 1;
        o = Math.floor(3 * Math.random());
        if (o == 0)
            o = 3;
        print("WANT TO MAKE A WAGER?");
        z = parseInt(await input());
        if (z != 0) {
            print("HOW MUCH ");
            while (1) {
                z1 = parseInt(await input());
                if (a1 < z1) {
                    print("TRIED TO FOOL ME; BET AGAIN");
                } else {
                    break;
                }
            }
        }
        w = 1;
        x = 1;
        y = 1;
        print("\n");
        print("IT'S YOUR MOVE:  ");
        while (1) {
            str = await input();
            p = parseInt(str);
            q = parseInt(str.substr(str.indexOf(",") + 1));
            r = parseInt(str.substr(str.lastIndexOf(",") + 1));
            if (p > w + 1 || q > x + 1 || r > y + 1 || (p == w + 1 && (q >= x + 1 || r >= y + 1)) || (q == x + 1 && r >= y + 1)) {
                print("\n");
                print("ILLEGAL MOVE, YOU LOSE.\n");
                break;
            }
            w = p;
            x = q;
            y = r;
            if (p == 3 && q == 3 && r == 3) {
                won = true;
                break;
            }
            if (p == a && q == b && r == c
             || p == d && q == e && r == f
             || p == g && q == h && r == i
             || p == j && q == k && r == l
             || p == m && q == n && r == o) {
                print("******BANG******");
                print("YOU LOSE!");
                print("\n");
                print("\n");
                won = false;
                break;
            }
            print("NEXT MOVE: ");
        }
        if (won) {
            print("CONGRATULATIONS!\n");
            if (z != 0) {
                z2 = a1 + z1;
                print("YOU NOW HAVE " + z2 + " DOLLARS.\n");
                a1 = z2;
            }
        } else {
            if (z != 0) {
                print("\n");
                z2 = a1 - z1;
                if (z2 <= 0) {
                    print("YOU BUST.\n");
                    break;
                } else {
                    print(" YOU NOW HAVE " + z2 + " DOLLARS.\n");
                    a1 = z2;
                }
            }
        }
        print("DO YOU WANT TO TRY AGAIN ");
        s = parseInt(await input());
        if (s != 1)
            break;
    }
    print("TOUGH LUCK!\n");
    print("\n");
    print("GOODBYE.\n");
}

main();

File: 30_Cube/java/src/Cube.java
import java.io.PrintStream;
import java.util.HashSet;
import java.util.Random;
import java.util.Scanner;
import java.util.Set;

/**
 * Game of Cube
 * <p>
 * Based on game of Cube at:
 * https://github.com/coding-horror/basic-computer-games/blob/main/30_Cube/cube.bas
 *
 *
 */
public class Cube {

    //Current player location
    private Location playerLocation;

    //Current list of mines
    private Set<Location> mines;

    //System input / output objects
    private PrintStream out;
    private Scanner scanner;

    //Player's current money
    private int money;

    /**
     * Entry point, creates a new Cube object and calls the play method
     * @param args Java execution arguments, not used in application
     */
    public static void main(String[] args) {
        new Cube().play();
    }

    public Cube() {
        out = System.out;
        scanner = new Scanner(System.in);
        money = 500;
        mines = new HashSet<>(5);
    }

    /**
     * Clears mines and places 5 new mines on the board
     */
    private void placeMines() {
        mines.clear();
        Random random = new Random();
        for(int i = 0; i < 5; i++) {
            int x = random.nextInt(1,4);
            int y = random.nextInt(1,4);
            int z = random.nextInt(1,4);
            mines.add(new Location(x,y,z));
        }
    }

    /**
     * Runs the entire game until the player runs out of money or chooses to stop
     */
    public void play() {
        out.println("DO YOU WANT TO SEE INSTRUCTIONS? (YES--1,NO--0)");
        if(readParsedBoolean()) {
            printInstructions();
        }
        do {
            placeMines();
            out.println("WANT TO MAKE A WAGER?");
            int wager = 0 ;

            if(readParsedBoolean()) {
                out.println("HOW MUCH?");
                do {
                    wager = Integer.parseInt(scanner.nextLine());
                    if(wager > money) {
                        out.println("TRIED TO FOOL ME; BET AGAIN");
                    }
                } while(wager > money);
            }

            playerLocation = new Location(1,1,1);
            while(playerLocation.x + playerLocation.y + playerLocation.z != 9) {
                out.println("\nNEXT MOVE");
                String input = scanner.nextLine();

                String[] stringValues = input.split(",");

                if(stringValues.length < 3) {
                    out.println("ILLEGAL MOVE, YOU LOSE.");
                    return;
                }

                int x = Integer.parseInt(stringValues[0]);
                int y = Integer.parseInt(stringValues[1]);
                int z = Integer.parseInt(stringValues[2]);

                Location location = new Location(x,y,z);

                if(x < 1 || x > 3 || y < 1 || y > 3 || z < 1 || z > 3 || !isMoveValid(playerLocation,location)) {
                    out.println("ILLEGAL MOVE, YOU LOSE.");
                    return;
                }

                playerLocation = location;

                if(mines.contains(location)) {
                    out.println("******BANG******");
                    out.println("YOU LOSE!\n\n");
                    money -= wager;
                    break;
                }
            }

            if(wager > 0) {
                out.printf("YOU NOW HAVE %d DOLLARS\n",money);
            }

        } while(money > 0 && doAnotherRound());

        out.println("TOUGH LUCK!");
        out.println("\nGOODBYE.");
    }

    /**
     * Queries the user whether they want to play another round
     * @return True if the player decides to play another round,
     * False if the player would not like to play again
     */
    private boolean doAnotherRound() {
        if(money > 0) {
            out.println("DO YOU WANT TO TRY AGAIN?");
            return readParsedBoolean();
        } else {
            return false;
        }
    }

    /**
     * Prints the instructions to the game, copied from the original code.
     */
    public void printInstructions() {
        out.println("THIS IS A GAME IN WHICH YOU WILL BE PLAYING AGAINST THE");
        out.println("RANDOM DECISION OF THE COMPUTER. THE FIELD OF PLAY IS A");
        out.println("CUBE OF SIDE 3. ANY OF THE 27 LOCATIONS CAN BE DESIGNATED");
        out.println("BY INPUTTING THREE NUMBERS SUCH AS 2,3,1. AT THE START");
        out.println("YOU ARE AUTOMATICALLY AT LOCATION 1,1,1. THE OBJECT OF");
        out.println("THE GAME IS TO GET TO LOCATION 3,3,3. ONE MINOR DETAIL:");
        out.println("THE COMPUTER WILL PICK, AT RANDOM, 5 LOCATIONS AT WHICH");
        out.println("IT WILL PLANT LAND MINES. IF YOU HIT ONE OF THESE LOCATIONS");
        out.println("YOU LOSE. ONE OTHER DETAIL: YOU MAY MOVE ONLY ONE SPACE");
        out.println("IN ONE DIRECTION EACH MOVE. FOR  EXAMPLE: FROM 1,1,2 YOU");
        out.println("MAY MOVE TO 2,1,2 OR 1,1,3. YOU MAY NOT CHANGE");
        out.println("TWO OF THE NUMBERS ON THE SAME MOVE. IF YOU MAKE AN ILLEGAL");
        out.println("MOVE, YOU LOSE AND THE COMPUTER TAKES THE MONEY YOU MAY");
        out.println("\n");
        out.println("ALL YES OR NO QUESTIONS WILL BE ANSWERED BY A 1 FOR YES");
        out.println("OR A 0 (ZERO) FOR NO.");
        out.println();
        out.println("WHEN STATING THE AMOUNT OF A WAGER, PRINT ONLY THE NUMBER");
        out.println("OF DOLLARS (EXAMPLE: 250)  YOU ARE AUTOMATICALLY STARTED WITH");
        out.println("500 DOLLARS IN YOUR ACCOUNT.");
        out.println();
        out.println("GOOD LUCK!");
    }

    /**
     * Waits for the user to input a boolean value. This could either be (true,false), (1,0), (y,n), (yes,no), etc.
     * By default, it will return false
     * @return Parsed boolean value of the user input
     */
    private boolean readParsedBoolean() {
        String in = scanner.nextLine();
        try {
            return in.toLowerCase().charAt(0) == 'y' || Boolean.parseBoolean(in) || Integer.parseInt(in) == 1;
        } catch(NumberFormatException exception) {
            return false;
        }
    }

    /**
     * Checks if a move is valid
     * @param from The point that the player is at
     * @param to The point that the player wishes to move to
     * @return True if the player is only moving, at most, 1 location in any direction, False if the move is invalid
     */
    private boolean isMoveValid(Location from, Location to) {
        return Math.abs(from.x - to.x) + Math.abs(from.y - to.y) + Math.abs(from.z - to.z) <= 1;
    }

    public class Location {
        int x,y,z;

        public Location(int x, int y, int z) {
            this.x = x;
            this.y = y;
            this.z = z;
        }

        /*
        For use in HashSet and checking if two Locations are the same
         */
        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Location location = (Location) o;

            if (x != location.x) return false;
            if (y != location.y) return false;
            return z == location.z;
        }

        /*
        For use in the HashSet to accordingly index the set
         */
        @Override
        public int hashCode() {
            int result = x;
            result = 31 * result + y;
            result = 31 * result + z;
            return result;
        }
    }
}
